docs/keps/20200619-kubefed-pull-reconciliation.md

```diff
@@ -0,0 +1,362 @@
+- [Proposal](#proposal)
+  - [Kubefed Daemon](#kubefed-daemon)
+  - [Kubefed Control-Plane](#kubefed-control-plane)
+  - [Controller Resource Propagation Status](#controller-resource-propagation-status)
+  - [Kubefed Security Concerns](#kubefed-security-concerns)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Design the new kubefed architecture to improve scalability and performance of
+kubefed on large scale scenarios.
+
+## Motivation
+
+All Kubefed logic is computed in the control-plane, that can represents a bottleneck
+whenever the amount of clusters and/or federated resources increases.
+The community has shared their intention to transition from a push-reconciler
+to a pull-reconciler approach.
+Scalability is important but also performance to avoid high response times when
+managing the lifecycle of certain federated resources.
+Additionally, users prefer to avoid giving write permissions of their clusters to
+store them in the control-plane cluster.
+
+### Goals
+
+* Design a pull-based reconciliation.
+* Scale kubefed to manage thousands of resources without major penalties.
+* Design a new architecture that could easily be extended to add new functionalities in the future.
+* Promote towards a more decentralized computation model.
+* The management of what is federated and where must remain centralized to the control-plane cluster.
+
+## Non Goals
+
+* Migrate all the controllers to use the controller-runtime. This intention exists
+but is not the ultimate goal during this refactoring.
+* Improve the status property of federated resources to reflect the current state of
+the resources and not the propagation status only.
+
+## Proposal
+
+This approach aims to split the current kubefed architecture into two parts: agents (aka kubefed daemons) and a control-plane.
+
+The kubefed agents are daemons running in the registered/joined kubefed clusters.
+The control-plane logic changes to use a pull-based reconciliation and rely on
+the kubefed daemons running on the target clusters to perform most of the computation.
+The federated resource management controllers remain in the cluster where the control-plane is deployed.
+
+<img src="./images/kubefedArch.jpg">
+
+### Kubefed Daemon
+
+On each registered kubefed cluster, a daemon should be deployed to watch the state
+of the cluster and federated resources. Likewise this daemon should periodically
+reconcile the desired state to create/delete/update the federated resources.
+
+As done in the past, the `kubefedctl join` command creates a kubefed cluster in
+the control-plane cluster, and now deploys the kubefed daemon in the target cluster.
+This operation also exchanges the required permissions (tokens, kubeconfigs) and url to enable
+the communication between the control-plane and the daemons, and viceversa.
+
+In the following we explain how this bi-directional communication occurs:
+
+* `control-plane --> kubefed daemon`: the control-plane polls the status of the cluster
```

Comment:
I see your concern here, and I have mix feelings. The idea was to follow the same approach for the kufebed daemons than the one used for the virtual kubelets. Additionally, the intention of exposing endpoints was to enable in the future endpoints to consume the logs of specific federated resources, e.g. `kubectl logs federateddeployment ...`.

However, we could follow an iterative approach here, and start using the current approach where the control plane queries these 2 details from the kube API. In a next iteration, we could expose these and other endpoints on the kubefed handlers. wdyt ?

Reply:
I understand the usage of exposing the enpoints; however I don't see the value (right away) justifying the amount of additional effort that needs to be put in. I see that we do concur on building this iteratively and lets update the doc to reflect the same theme.

Reply:
I also feel that exposing custom endpoints for resource status is less ideal than talking to the managed cluster api server directly. Even for log/exec endpoints, I guess you could leverage the api server and treat it as a proxy?

I actually prefer a *fully* decentralized model where the kubefed daemon will just update federated resource status in control plane API server directly, instead of this mixed model. IIUC, the main goal here is to address the scalability issue. I understand the security risk of exposing write permission to kubefed daemon in that solution. The question would be: can we leverage RBAC in control plane cluster to restrict the write access only to specific status field of some federated resources?

Reply:
Okay, I like @jieyu approach where the propagation and reconciliation would be now decentralized. I also see you two have concerns about the endpoints approach due to the maintenance cost. I'd change this initial approach and transition to this decentralized model. I'd try to think about any potential corner case.

I believe we could control the read/write access of kubefed clusters to theecontrol plane cluster via webhooks and RBAC rules, but I am not 100% sure about any other considerations.

Reply:
I personally prefer a shim API so that a cluster only sees resources and fields that are related to it. For example, if there are overrides specific to certain clusters, I would not want any cluster to be able to see them but rather only the override applicable to the specific cluster.